src/providers/openstack/configdrive.rs

```diff
@@ -0,0 +1,273 @@
+//! configdrive metadata fetcher for OpenStack
+//! reference: https://docs.openstack.org/nova/latest/user/metadata.html
+
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{BufReader, Read};
+use std::path::{Path, PathBuf};
+
+use openssh_keys::PublicKey;
+use slog_scope::{error, warn};
+use tempfile::TempDir;
+
+use crate::errors::*;
+use crate::network;
+use crate::providers::MetadataProvider;
+
+const CONFIG_DRIVE_LABEL: &str = "config-2";
+
+/// Partial object for ec2 `meta_data.json`
+#[derive(Debug, Deserialize)]
+pub struct MetadataEc2JSON {
+    /// Local hostname.
+    pub hostname: Option<String>,
```

Comment:
I think you can drop this from here as this detail is now coming from `MetadataOpenstackJSON`.